extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,202 @@
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = !!(this.config_  // The network has opted-in.
+        // The publisher has enabled refresh on this slot.
+        && (this.refreshInterval_ || this.refreshInterval_ == '')
+        // The slot is contained only within container types eligible for
+        // refresh.
+        && !getEnclosingContainerTypes(this.element_).filter(container =>
+          container != ValidAdContainerTypes['AMP-CAROUSEL']
+          && container != ValidAdContainerTypes['AMP-STICKY-AD']).length);
+  }
+
+
+  /**
+   * Initiates the refresh cycle by initiating the visibility manager on the
+   * element.
+   *
+   * @return {?Promise} Promise that resolves after refresh event. This promise
+   *   is useful to have for test purposes.
+   */
+  initiateRefreshCycle() {
+    if (!this.isRefreshable()) {
+      return null;
+    }
+    return new Promise(resolve => {
+      analyticsForDoc(this.element_, true).then(analytics => {
+        analytics.getAnalyticsRoot(this.element_).getVisibilityManager()
+            .listenElement(this.element_, this.config_, null, null, () => {
+              this.refreshTimeoutId_ = this.timer_.delay(() => {
+                this.a4a_.refresh(() => this.initiateRefreshCycle());
+                resolve();
+              }, /** @type {number} */ (this.refreshInterval_));
+            });
+      });
+    });
+  }
+
+  /**
+   * Terminates the current refresh cycle, if one currently exists. Returns
+   * true if the cycle was canceled successfully, false otherwise.
+   */
+  stopRefreshCycle() {
+    if (this.refreshTimeoutId_) {
+      this.timer_.cancel(this.refreshTimeoutId_);
+      this.refreshTimeoutId_ = null;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Converts config to appropriate units.
+   * @return {!RefreshConfig}
+   */
+  getConfiguration_(config) {
+    // Convert seconds to milliseconds.
+    config['totalTimeMin'] *= 1000;
+    config['continuousTimeMin'] *= 1000;
+    return config;
+  }
+
+  /**
+   * Retrieves the publisher-specified refresh interval, if one were set.
+   *
+   * @return {?number}
+   */
+  getPublisherSpecifiedRefreshInterval_() {
+    const refreshInterval = this.element_.getAttribute(DATA_ATTR_NAME);
+    if (refreshInterval) {
+      return this.checkAndSanitizeRefreshInterval_(refreshInterval);
+    }
+    let metaTag;
+    const metaTagContent = ((metaTag = this.win_.document
+          .getElementsByName(METATAG_NAME))
+        && metaTag[0]
+        && metaTag[0].getAttribute('content'));
+    const networkIntervalPairs =
+        metaTagContent ? metaTagContent.split(',') : [];
+    for (let i = 0; i < networkIntervalPairs.length; i++) {
+      const pair = networkIntervalPairs[i].split('=');
+      if (pair.length != 2) {
```

Comment:
Can this just be:

user().assert(pair.length == 2, 'refresh metadata...');
if (pair[0] == this.adType_) {
    return ...
}